File: lista python/at5.py
class Animal:
    def __init__(self, nome):
        self.nome = nome
    
    def som (self):
        raise NotImplementedError("Subclasses devem implementar o método som.")
    

class Cachorro(Animal):
    def som(self):
        return f"{self.nome}: Au au!"
    

class Gato(Animal):
     def som(self):
        return f"{self.nome}: Miau!"
     


def emitir_sons(animais):
    for animal in animais:
        print(animal.som())

cachorro = Cachorro("Mike")
gato = Gato("Garfield")

lista_de_animais = [cachorro, gato]


emitir_sons(lista_de_animais)

File: lista java/at5.java
import java.util.ArrayList;
import java.util.List;


class Animal {
    public void som() {
        System.out.println("O animal faz um som.");
    }
}


class Cachorro extends Animal {
    @Override
    public void som() {
        System.out.println("O cachorro late: Au Au!");
    }
}


class Gato extends Animal {
    @Override
    public void som() {
        System.out.println("O gato mia: Miau!");
    }
}

public class Main {

    public static void emitirSonsDosAnimais(List<Animal> animais) {
        for (Animal animal : animais) {
            animal.som();  
        }
    }

    public static void main(String[] args) {
        
        List<Animal> animais = new ArrayList<>();
        animais.add(new Cachorro());
        animais.add(new Gato());

        
        emitirSonsDosAnimais(animais);
    }
}
